accounts-db/benches/read_only_accounts_cache.rs

```diff
@@ -0,0 +1,270 @@
+#![feature(test)]
+
+extern crate test;
+
+use {
+    criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion},
+    rand::{rngs::SmallRng, seq::SliceRandom, SeedableRng},
+    solana_accounts_db::{
+        accounts_db::AccountsDb,
+        read_only_accounts_cache::{ReadOnlyAccountsCache, CACHE_ENTRY_SIZE},
+    },
+    solana_sdk::{
+        account::{Account, ReadableAccount},
+        pubkey,
+    },
+    std::{
+        iter,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread::Builder,
+    },
+};
+
+/// Sizes of accounts to bench.
+const ACCOUNTS_SIZES: &[usize] = &[0, 512, 1024];
+/// Numbers of reader and writer threads to bench.
+const NUM_READERS_WRITERS: &[usize] = &[
+    8,
+    16,
+    // These parameters are likely to freeze your computer, if it has less than
+    // 32 cores.
+    // 32, 64, 128, 256, 512, 1024,
+];
+
+fn bench_read_only_accounts_cache(c: &mut Criterion) {
+    let mut group = c.benchmark_group("cache");
+    let slot = 0;
+
+    for account_size in ACCOUNTS_SIZES {
+        // Number of accounts to use in the benchmark.
+        let num_accounts_benched = AccountsDb::DEFAULT_MAX_READ_ONLY_CACHE_DATA_SIZE_HI
+            .div_ceil(CACHE_ENTRY_SIZE.saturating_add(*account_size));
+        group.sample_size(num_accounts_benched);
+
+        for num_readers_writers in NUM_READERS_WRITERS {
+            let cache = Arc::new(ReadOnlyAccountsCache::new(
+                AccountsDb::DEFAULT_MAX_READ_ONLY_CACHE_DATA_SIZE_LO,
+                AccountsDb::DEFAULT_MAX_READ_ONLY_CACHE_DATA_SIZE_HI,
+                AccountsDb::READ_ONLY_CACHE_MS_TO_SKIP_LRU_UPDATE,
+            ));
+
+            // Benchmark the performance of loading and storing accounts in an
+            // initially empty cache.
+            let pubkeys: Vec<_> = iter::repeat_with(pubkey::new_rand)
+                .take(num_accounts_benched)
+                .collect();
+            group.bench_function(
+                BenchmarkId::new(
+                    "read_only_accounts_cache_store",
+                    format!("{account_size}_bytes_{num_readers_writers}_threads"),
+                ),
+                |b| {
+                    b.iter_batched(
+                        || {
+                            let accounts_data = iter::repeat(
+                                Account {
+                                    lamports: 1,
+                                    data: vec![1; *account_size],
```

Comment:
You should probably create these accounts before starting the bench - setup is
still executed at each iteration. Then just clone - cloning AccountSharedData is
cheap it's just an Arc clone